src/Kestrel.Core/Internal/Http2/Http2Connection.cs

```diff
@@ -988,6 +1027,24 @@ void IHttp2StreamLifetimeHandler.OnStreamCompleted(int streamId)
             }
         }
 
+        private void StartKeepAliveTimeout()
```

Comment:
Looks like these methods should only be called under the _stateLock. We should add a comment just to remind ourselves.

Reply:
Actually seems like the only contention would be _currentTimeout having not updated yet. Why not expose the _timeoutReason on TimeoutControl?

Reply:
I like adding a CurrentSetTimeoutReason property to TimeoutControl, but it definitely won't remove the locking requirement.